rdflib/graph.py

```diff
@@ -1070,13 +1073,24 @@ def parse(
         )
         if format is None:
             format = source.content_type
+        assumed_xml = False
         if format is None:
-            # raise Exception("Could not determine format for %r. You can" + \
-            # "expicitly specify one with the format argument." % source)
-            format = "application/rdf+xml"
+            try:
+                format = rdflib.util.guess_format(source.file.name)
```

Comment:
Under what circumstances are you seeing these errors? I'm guessing when `source.file.name is None` for the TypeError and probably ... when `not hasattr(source, 'file')`? If that is the case then an explicit test (if statement) for the presence and non-Noneness of the values would be much clearer here.

Reply:
@tgbugs I've adopted your suggestion for more of a look-before-you-leap approach with an explicit if-statement guard, ensuring that `source.file.name` is an accessible path and is a string before attempting a call to `guess_format`.